### PR TITLE
ES6 conversion

### PR DESCRIPTION
**Pull request by rettgerst**
I updated the source code to use ES6 classes instead of old style function declarations/prototype assignment, as well as converting the source code to TypeScript.

I also added some basic project setup including a package.json, plus some tooling including prettier and eslint.

Beyond the ES6 class refactor I didn't do much modification to the original source code. There are a lot of TS and ESLint errors. However, the code builds and runs! I tested it by playing Sonic Advance 3 for a few minutes and all works as expected.

### CODE REVIEW

**Comment by andychase** [comment]
Would you be opposed to stopping at the es6 and not implementing typescript yet?

Although the project would probably benefit from full typing, right now it's a fully pure javascript emulator and I think there's some value in keeping it that way.

**Comment by rettgerst** [comment]
I removed the typescript conversion commit from my branch so now it's all vanilla javascript, although it seems the upstream has changed yesterday so I will need to resolve that as well when I get the time to.

**Comment by rettgerst** [comment]
okay, synced with upstream now.

**Comment by andychase** [comment]
Reviewing this now

**Comment by andychase** [comment]
Didn't work when I tried it, got error message:

```
gba.js:86 Uncaught TypeError: Failed to execute 'drawImage' on 'CanvasRenderingContext2D': The provided value is not of type '(CSSImageValue or HTMLImageElement or SVGImageElement or HTMLVideoElement or HTMLCanvasElement or ImageBitmap or OffscreenCanvas)'
    at GameBoyAdvanceVideo.video.drawCallback (gba.js:86)
    at GameBoyAdvanceVideo.finishDraw (video.js:182)
    at Object.finish (proxy.js:89)
    at Worker.GameBoyAdvanceRenderProxy.worker.onmessage (proxy.js:94)
```

I'll see if I can easily fix the problem

File: js/gba.js [hunk 1]
```diff
@@ -1,418 +1,404 @@
...
+		if (canvas.offsetWidth != 240 || canvas.offsetHeight != 160) {
+			this.indirectCanvas = document.createElement("canvas");
+			this.indirectCanvas.setAttribute("height", "160");
+			this.indirectCanvas.setAttribute("width", "240");
+			this.targetCanvas = canvas;
+			this.setCanvasDirect(this.indirectCanvas);
+			var targetContext = canvas.getContext("2d");
+			this.video.drawCallback = function () {
+				targetContext.drawImage(
+					this.indirectCanvas,
```
Comment (andychase) [inline js/gba.js:87]: the `this`. instead of `self.` is what's causing the emulator not to work.

File: js/video/software.js [hunk 2]
```diff
@@ -1,1606 +1,1657 @@
...
-	}
-
-	MemoryAligned16.prototype.store16.call(this, offset, value);
-};
-
-function GameBoyAdvancePalette() {
-	this.colors = [ new Array(0x100), new Array(0x100) ];
-	this.adjustedColors = [ new Array(0x100), new Array(0x100) ];
-	this.passthroughColors = [
-		this.colors[0], // BG0
```
Comment (andychase) [inline js/video/software.js:168]: Why where these comments removed? I think they are helpful

Comment (rettgerst) [reply js/video/software.js:168]: the refactoring to ES6 was done automatically by VSCode which must have automatically removed the comments, I can re-add them.

Comment (andychase) [reply js/video/software.js:168]: Ok no problem. That's odd that it removed some but not others. Actually give me a few minutes I'm about to make a PR to your branch where I'll re-add them

**Comment by andychase** [comment]
First of all thank you so much for doing the work to convert to ES6. I do appreciate it.

I opened a PR to your branch here:

https://github.com/rettgerst/gbajs2/pull/1

* Some bit twiddling code looked better on the old version so I reverted those. May be we will have to mark those parts to be ignored by reformatting tools?
* The hex values are all lowercase.. It's fine for now but probably will want to re-uppercase them since they are used heavily. Unfortunately prettier doesn't support this at all so it's tough.

I reviewed it and everything looks good after my changes

**Comment by rettgerst** [comment]
no problem. prettier is what I use at work so it's what I went with in the absence of a formatter, but you can use a different one or nothing at all if you prefer.

**Pull request merged**